README.md
# Number Length

[![Hackage](http://img.shields.io/hackage/v/number-length.svg)][Hackage: number-length]
[![Hackage Dependencies](https://img.shields.io/hackage-deps/v/number-length.svg)](http://packdeps.haskellers.com/reverse/number-length)
[![Haskell Programming Language](https://img.shields.io/badge/language-Haskell-blue.svg)][Haskell.org]
[![BSD3 License](http://img.shields.io/badge/license-BSD3-brightgreen.svg)][tl;dr Legal: BSD3]

[![Build](https://travis-ci.org/trskop/number-length.svg)](https://travis-ci.org/trskop/number-length)
[![Coverage Status](https://coveralls.io/repos/github/trskop/number-length/badge.svg?branch=master)](https://coveralls.io/github/trskop/number-length?branch=master)


## Description

Number length in decimal and hexadecimal representation.

There are situations when it is necessary to know the length of a number in
decimal or hexadecimal form. In example when allocating buffer while
serializing a number in to binary or human readable format.

    λ> numberLength (123 :: Int)
    3

    λ> numberLengthHex (123 :: Int)  -- 123 = 7b in hex
    2


## Building Options

* `-fpedantic` (disabled by default)

  Pass additional warning flags to GHC.


## License

The BSD 3-Clause License, see [LICENSE][] file for details.


## Contributions

Contributions, pull requests and bug reports are welcome! Please don't be
afraid to contact author using GitHub or by e-mail.


[Hackage: number-length]:
  http://hackage.haskell.org/package/number-length
  "Package number-length on Hackage"
[Haskell.org]:
  http://www.haskell.org
  "The Haskell Programming Language"
[LICENSE]:
  https://github.com/trskop/number-length/blob/master/LICENSE
  "License of number-length package."
[tl;dr Legal: BSD3]:
  https://tldrlegal.com/license/bsd-3-clause-license-%28revised%29
  "BSD 3-Clause License (Revised)"


ChangeLog.md
# ChangeLog / ReleaseNotes


## Version 0.2.1.0

* Depend on `quickcheck-instances` to provide `instance Arbitrary Natural` for
  unit tests. (**minor change**)


## Version 0.2.0.0

* Instances for `Integer` and `Natural`.
* Build on GHC 8.2.2 and Stackage LTS 10.\*.


## Version 0.1.0.1

* Removed dead code. (**minor change**)
* Better and expanded test suite, which now includes tests for methods of
  `SignedNumberLength`. (**minor change**)
* Depending on tagged package only for GHC <7.8. This is due to the fact, that
  `Data.Proxy` module was introduce in base ==4.7, which was bundled with GHC
  7.8. (**minor change**)
* Small examples included in description. (**minor change**)
* Uploaded to [Hackage][]:
  <http://hackage.haskell.org/package/number-length-0.1.0.1>


## Version 0.1.0.0

* First public release.
* Uploaded to [Hackage][]:
  <http://hackage.haskell.org/package/number-length-0.1.0.0>



[Hackage]:
  http://hackage.haskell.org/
  "HackageDB (or just Hackage) is a collection of releases of Haskell packages."


test/TestCase.hs
{-# LANGUAGE CPP #-}
{-# LANGUAGE NoImplicitPrelude #-}
-- |
-- Module:       $HEADER$
-- Description:  All test cases aggregated and exported as tests :: [Test].
-- Copyright:    (c) 2015, Peter Trško
-- License:      BSD3
--
-- Stability:    stable
-- Portability:  NoImplicitPrelude
--
-- All test cases aggregated and exported as @tests :: ['Test']@.
module TestCase (tests)
    where

import Test.Framework (Test, testGroup)

import qualified TestCase.Data.NumberLength as NumberLength (tests)
import qualified TestCase.Data.NumberLength.Int as Int (tests)
import qualified TestCase.Data.NumberLength.Integer as Integer (tests)
import qualified TestCase.Data.NumberLength.Word as Word (tests)

#ifdef HAVE_NATURAL
import qualified TestCase.Data.NumberLength.Natural as Natural (tests)
#endif


tests :: [Test]
tests =
    [ testGroup "Data.NumberLength.Int" Int.tests
    , testGroup "Data.NumberLength.Word" Word.tests
    , testGroup "Data.NumberLength.Integer" Integer.tests
#ifdef HAVE_NATURAL
    , testGroup "Data.NumberLength.Natural" Natural.tests
#endif
    , testGroup "Data.NumberLength" NumberLength.tests
    ]


test/benchmark.hs
{-# LANGUAGE CPP #-}
{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE NoMonomorphismRestriction #-}
-- |
-- Module:       Main
-- Description:  Benchmarks for low-level functions.
-- Copyright:    (c) 2016, Peter Trško
-- License:      BSD3
--
-- Stability:    stable
-- Portability:  NoImplicitPrelude
--
-- Benchmarks for low-level functions.
module Main (main)
  where

import Prelude (Bounded(minBound, maxBound), Num((*)), fromIntegral)

import Data.Function (($))
import Data.Int (Int64)
import Data.Word (Word64)
import System.IO (IO)

import Criterion.Main (bench, defaultMain, nf)

import Data.NumberLength.Int
    ( lengthInt
    , lengthInt16
    , lengthInt16hex
    , lengthInt32
    , lengthInt32hex
    , lengthInt64
    , lengthInt64hex
    , lengthInt8
    , lengthInt8hex
    , lengthIntHex
    )
import Data.NumberLength.Integer
    ( lengthInteger
    , lengthIntegerHex
    )
#ifdef HAVE_NATURAL
import Data.NumberLength.Natural
    ( lengthNatural
    , lengthNaturalHex
    )
#endif
import Data.NumberLength.Word
    ( lengthWord
    , lengthWord16
    , lengthWord16hex
    , lengthWord32
    , lengthWord32hex
    , lengthWord64
    , lengthWord64hex
    , lengthWord8
    , lengthWord8hex
    , lengthWordHex
    )


main :: IO ()
main = defaultMain
    [ bench "lengthInt minBound" $ nf lengthInt minBound
    , bench "lengthInt 0"        $ nf lengthInt 0
    , bench "lengthInt maxBound" $ nf lengthInt maxBound

    , bench "lengthInt8 minBound" $ nf lengthInt8 minBound
    , bench "lengthInt8 0"        $ nf lengthInt8 0
    , bench "lengthInt8 maxBound" $ nf lengthInt8 maxBound

    , bench "lengthInt16 minBound" $ nf lengthInt16 minBound
    , bench "lengthInt16 0"        $ nf lengthInt16 0
    , bench "lengthInt16 maxBound" $ nf lengthInt16 maxBound

    , bench "lengthInt32 minBound" $ nf lengthInt32 minBound
    , bench "lengthInt32 0"        $ nf lengthInt32 0
    , bench "lengthInt32 maxBound" $ nf lengthInt32 maxBound

    , bench "lengthInt64 minBound" $ nf lengthInt64 minBound
    , bench "lengthInt64 0"        $ nf lengthInt64 0
    , bench "lengthInt64 maxBound" $ nf lengthInt64 maxBound

    , bench "lengthIntHex minBound" $ nf lengthIntHex minBound
    , bench "lengthIntHex 0"        $ nf lengthIntHex 0
    , bench "lengthIntHex maxBound" $ nf lengthIntHex maxBound

    , bench "lengthInt8hex minBound" $ nf lengthInt8hex minBound
    , bench "lengthInt8hex 0"        $ nf lengthInt8hex 0
    , bench "lengthInt8hex maxBound" $ nf lengthInt8hex maxBound

    , bench "lengthInt16hex minBound" $ nf lengthInt16hex minBound
    , bench "lengthInt16hex 0"        $ nf lengthInt16hex 0
    , bench "lengthInt16hex maxBound" $ nf lengthInt16hex maxBound

    , bench "lengthInt32hex minBound" $ nf lengthInt32hex minBound
    , bench "lengthInt32hex 0"        $ nf lengthInt32hex 0
    , bench "lengthInt32hex maxBound" $ nf lengthInt32hex maxBound

    , bench "lengthInt64hex minBound" $ nf lengthInt64hex minBound
    , bench "lengthInt64hex 0"        $ nf lengthInt64hex 0
    , bench "lengthInt64hex maxBound" $ nf lengthInt64hex maxBound


    , bench "lengthInteger minInt64"           $ nf lengthInteger minInt64
    , bench "lengthInteger 0"                  $ nf lengthInteger 0
    , bench "lengthInteger maxInt64"           $ nf lengthInteger maxInt64
    , bench "lengthInteger maxWord64"          $ nf lengthInteger maxWord64
    , bench "lengthInteger (maxWord64 * 2)"    $ nf lengthInteger maxWord64x2

    , bench "lengthIntegerHex minInt64"        $ nf lengthIntegerHex minInt64
    , bench "lengthIntegerHex 0"               $ nf lengthIntegerHex 0
    , bench "lengthIntegerHex minInt64"        $ nf lengthIntegerHex maxInt64
    , bench "lengthIntegerHex maxWord64"       $ nf lengthIntegerHex maxWord64
    , bench "lengthIntegerHex (maxWord64 * 2)" $ nf lengthIntegerHex maxWord64x2


#ifdef HAVE_NATURAL
    , bench "lengthNatural 0"                  $ nf lengthNatural 0
    , bench "lengthNatural maxInt64"           $ nf lengthNatural maxInt64
    , bench "lengthNatural maxWord64"          $ nf lengthNatural maxInt64
    , bench "lengthNatural (maxWord64 * 2)"    $ nf lengthNatural maxWord64x2

    , bench "lengthNaturalHex 0"               $ nf lengthNaturalHex 0
    , bench "lengthNaturalHex minInt64"        $ nf lengthNaturalHex maxInt64
    , bench "lengthNaturalHex maxWord64"       $ nf lengthNaturalHex maxWord64
    , bench "lengthNaturalHex (maxWord64 * 2)" $ nf lengthNaturalHex maxWord64x2
#endif


    , bench "lengthWord 0"        $ nf lengthWord 0
    , bench "lengthWord maxBound" $ nf lengthWord maxBound

    , bench "lengthWord8 0"        $ nf lengthWord8 0
    , bench "lengthWord8 maxBound" $ nf lengthWord8 maxBound

    , bench "lengthWord16 0"        $ nf lengthWord16 0
    , bench "lengthWord16 maxBound" $ nf lengthWord16 maxBound

    , bench "lengthWord32 0"        $ nf lengthWord32 0
    , bench "lengthWord32 maxBound" $ nf lengthWord32 maxBound

    , bench "lengthWord64 0"        $ nf lengthWord64 0
    , bench "lengthWord64 maxBound" $ nf lengthWord64 maxBound

    , bench "lengthWordHex 0"        $ nf lengthWordHex 0
    , bench "lengthWordHex maxBound" $ nf lengthWordHex maxBound

    , bench "lengthWord8hex 0"        $ nf lengthWord8hex 0
    , bench "lengthWord8hex maxBound" $ nf lengthWord8hex maxBound

    , bench "lengthWord16hex 0"        $ nf lengthWord16hex 0
    , bench "lengthWord16hex maxBound" $ nf lengthWord16hex maxBound

    , bench "lengthWord32hex 0"        $ nf lengthWord32hex 0
    , bench "lengthWord32hex maxBound" $ nf lengthWord32hex maxBound

    , bench "lengthWord64hex 0"        $ nf lengthWord64hex 0
    , bench "lengthWord64hex maxBound" $ nf lengthWord64hex maxBound
    ]
  where
    maxInt64 = fromIntegral (maxBound :: Int64)
    minInt64 = fromIntegral (minBound :: Int64)
    maxWord64 = fromIntegral (maxBound :: Word64)
    maxWord64x2 = maxWord64 * 2


test/TestCase/Data/Int/Length.hs
{-# LANGUAGE NoImplicitPrelude #-}
-- |
-- Module:       $HEADER$
-- Description:  TODO
-- Copyright:    (c) 2015, Peter Trško
-- License:      BSD3
--
-- Stability:    experimental
-- Portability:  NoImplicitPrelude
module TestCase.Data.Int.Length
  where

import Prelude
    ( Bounded(maxBound, minBound)
    , Integral
    , Num(negate)
    , fromIntegral
    )

import Control.Applicative (liftA2)
import Data.Bool (Bool, otherwise)
import Data.Eq (Eq((==)))
import Data.Function ((.), ($), id)
import Data.Int (Int, Int16, Int32, Int64, Int8)
import qualified Data.List as List (dropWhile, length)
import Data.Ord (Ord((<)))
import Data.String (String)
import Text.Printf (PrintfArg, printf)
import Text.Show (Show(show))

import Test.HUnit (Assertion, (@?=))
import Test.Framework (Test)
import Test.Framework.Providers.HUnit (testCase)
import Test.Framework.Providers.QuickCheck2 (testProperty)

import Data.Int.Length
    ( lengthInt
    , lengthInt16
    , lengthInt16hex
    , lengthInt32
    , lengthInt32hex
    , lengthInt64
    , lengthInt64hex
    , lengthInt8
    , lengthInt8hex
    , lengthIntHex
    )


tests :: [Test]
tests =
    [ testCase     "lengthInt8 minBound"         test_lengthInt8_minBound
    , testCase     "lengthInt8 maxBound"         test_lengthInt8_maxBound
    , testProperty "lengthInt8 = length . show"  property_lengthInt8

    , testCase     "lengthInt16 minBound"        test_lengthInt16_minBound
    , testCase     "lengthInt16 maxBound"        test_lengthInt16_maxBound
    , testProperty "lengthInt16 = length . show" property_lengthInt16

    , testCase     "lengthInt32 minBound"        test_lengthInt32_minBound
    , testCase     "lengthInt32 maxBound"        test_lengthInt32_maxBound
    , testProperty "lengthInt32 = length . show" property_lengthInt32

    , testCase     "lengthInt64 minBound"        test_lengthInt64_minBound
    , testCase     "lengthInt64 maxBound"        test_lengthInt64_maxBound
    , testProperty "lengthInt64 = length . show" property_lengthInt64

    , testCase     "lengthInt minBound"          test_lengthInt_minBound
    , testCase     "lengthInt maxBound"          test_lengthInt_maxBound
    , testProperty "lengthInt = length . show"   property_lengthInt

    , testCase     "lengthInt8hex minBound"         test_lengthInt8hex_minBound
    , testCase     "lengthInt8hex maxBound"         test_lengthInt8hex_maxBound
    , testProperty "lengthInt8hex = length . show"  property_lengthInt8hex

    , testCase     "lengthInt16hex minBound"        test_lengthInt16hex_minBound
    , testCase     "lengthInt16hex maxBound"        test_lengthInt16hex_maxBound
    , testProperty "lengthInt16hex = length . show" property_lengthInt16hex

    , testCase     "lengthInt32hex minBound"        test_lengthInt32hex_minBound
    , testCase     "lengthInt32hex maxBound"        test_lengthInt32hex_maxBound
    , testProperty "lengthInt32hex = length . show" property_lengthInt32hex

    , testCase     "lengthInt64hex minBound"        test_lengthInt64hex_minBound
    , testCase     "lengthInt64hex maxBound"        test_lengthInt64hex_maxBound
    , testProperty "lengthInt64hex = length . show" property_lengthInt64hex

    , testCase     "lengthIntHex minBound"          test_lengthIntHex_minBound
    , testCase     "lengthIntHex maxBound"          test_lengthIntHex_maxBound
    , testProperty "lengthIntHex = length . show"   property_lengthIntHex
    ]

numberLengthDec :: Show n => n -> Int
numberLengthDec = List.length . List.dropWhile (== '-') . show

numberLengthHex :: (PrintfArg n, Integral n, Ord n) => n -> Int
numberLengthHex n
  | n < 0     = numberLengthHex' . negate . int64 $ fromIntegral n
  | otherwise = numberLengthHex' n
  where
    numberLengthHex' :: PrintfArg n => n -> Int
    numberLengthHex' m = List.length (printf "%x" m :: String)

    int64 :: Int64 -> Int64
    int64 = id

(<==>) :: Eq b => (a -> b) -> (a -> b) -> a -> Bool
(<==>) = liftA2 (==)

-- {{{ Decimal ----------------------------------------------------------------
-- {{{ lengthInt8 -------------------------------------------------------------

test_lengthInt8_minBound, test_lengthInt8_maxBound :: Assertion

test_lengthInt8_minBound =
    lengthInt8 minBound @?= numberLengthDec (minBound :: Int8)

test_lengthInt8_maxBound =
    lengthInt8 maxBound @?= numberLengthDec (maxBound :: Int8)

property_lengthInt8 :: Int8 -> Bool
property_lengthInt8 = lengthInt8 <==> numberLengthDec

-- }}} lengthInt8 -------------------------------------------------------------
-- {{{ lengthInt16 ------------------------------------------------------------

test_lengthInt16_minBound, test_lengthInt16_maxBound :: Assertion

test_lengthInt16_minBound =
    lengthInt16 minBound @?= numberLengthDec (minBound :: Int16)

test_lengthInt16_maxBound =
    lengthInt16 maxBound @?= numberLengthDec (maxBound :: Int16)

property_lengthInt16 :: Int16 -> Bool
property_lengthInt16 = lengthInt16 <==> numberLengthDec

-- }}} lengthInt16 ------------------------------------------------------------
-- {{{ lengthInt32 ------------------------------------------------------------

test_lengthInt32_minBound, test_lengthInt32_maxBound :: Assertion

test_lengthInt32_minBound =
    lengthInt32 minBound @?= numberLengthDec (minBound :: Int32)

test_lengthInt32_maxBound =
    lengthInt32 maxBound @?= numberLengthDec (maxBound :: Int32)

property_lengthInt32 :: Int32 -> Bool
property_lengthInt32 = lengthInt32 <==> numberLengthDec

-- }}} lengthInt32 ------------------------------------------------------------
-- {{{ lengthInt64 ------------------------------------------------------------

test_lengthInt64_minBound, test_lengthInt64_maxBound :: Assertion

test_lengthInt64_minBound =
    lengthInt64 minBound @?= numberLengthDec (minBound :: Int64)

test_lengthInt64_maxBound =
    lengthInt64 maxBound @?= numberLengthDec (maxBound :: Int64)

property_lengthInt64 :: Int64 -> Bool
property_lengthInt64 = lengthInt64 <==> numberLengthDec

-- }}} lengthInt64 ------------------------------------------------------------
-- {{{ lengthInt --------------------------------------------------------------

test_lengthInt_minBound, test_lengthInt_maxBound :: Assertion

test_lengthInt_minBound =
    lengthInt minBound @?= numberLengthDec (minBound :: Int)

test_lengthInt_maxBound =
    lengthInt maxBound @?= numberLengthDec (maxBound :: Int)

property_lengthInt :: Int -> Bool
property_lengthInt = lengthInt <==> numberLengthDec

-- }}} lengthInt --------------------------------------------------------------
-- }}} Decimal ----------------------------------------------------------------

-- {{{ Hexadecimal ------------------------------------------------------------
-- {{{ lengthInt8hex ----------------------------------------------------------

test_lengthInt8hex_minBound, test_lengthInt8hex_maxBound :: Assertion

test_lengthInt8hex_minBound =
    lengthInt8hex minBound @?= numberLengthHex (minBound :: Int8)

test_lengthInt8hex_maxBound =
    lengthInt8hex maxBound @?= numberLengthHex (maxBound :: Int8)

property_lengthInt8hex :: Int8 -> Bool
property_lengthInt8hex = lengthInt8hex <==> numberLengthHex

-- }}} lengthInt8hex ----------------------------------------------------------
-- {{{ lengthInt16hex ---------------------------------------------------------

test_lengthInt16hex_minBound, test_lengthInt16hex_maxBound :: Assertion

test_lengthInt16hex_minBound =
    lengthInt16hex minBound @?= numberLengthHex (minBound :: Int16)

test_lengthInt16hex_maxBound =
    lengthInt16hex maxBound @?= numberLengthHex (maxBound :: Int16)

property_lengthInt16hex :: Int16 -> Bool
property_lengthInt16hex = lengthInt16hex <==> numberLengthHex

-- }}} lengthInt16hex ---------------------------------------------------------
-- {{{ lengthInt32hex ---------------------------------------------------------

test_lengthInt32hex_minBound, test_lengthInt32hex_maxBound :: Assertion

test_lengthInt32hex_minBound =
    lengthInt32hex minBound @?= numberLengthHex (minBound :: Int32)

test_lengthInt32hex_maxBound =
    lengthInt32hex maxBound @?= numberLengthHex (maxBound :: Int32)

property_lengthInt32hex :: Int32 -> Bool
property_lengthInt32hex = lengthInt32hex <==> numberLengthHex

-- }}} lengthInt32hex ---------------------------------------------------------
-- {{{ lengthInt64hex ---------------------------------------------------------

test_lengthInt64hex_minBound, test_lengthInt64hex_maxBound :: Assertion

test_lengthInt64hex_minBound =
    lengthInt64hex minBound @?= 16 -- negate issue

test_lengthInt64hex_maxBound =
    lengthInt64hex maxBound @?= numberLengthHex (maxBound :: Int64)

property_lengthInt64hex :: Int64 -> Bool
property_lengthInt64hex = lengthInt64hex <==> numberLengthHex

-- }}} lengthInt64hex ---------------------------------------------------------
-- {{{ lengthIntHex -----------------------------------------------------------

test_lengthIntHex_minBound, test_lengthIntHex_maxBound :: Assertion

test_lengthIntHex_minBound = lengthIntHex minBound @?=
    -- Handling of "negate" issue:
    if (minBound :: Int) == fromIntegral (minBound :: Int32)
        then 8  -- 32bit
        else 16 -- 64bit

test_lengthIntHex_maxBound =
    lengthIntHex maxBound @?= numberLengthHex (maxBound :: Int)

property_lengthIntHex :: Int -> Bool
property_lengthIntHex = lengthIntHex <==> numberLengthHex

-- }}} lengthIntHex -----------------------------------------------------------
-- }}} Hexadecimal ------------------------------------------------------------


src/Data/Word/Length.hs
{-# LANGUAGE NoImplicitPrelude #-}
-- |
-- Module:       $HEADER$
-- Description:  TODO
-- Copyright:    (c) 2015, Peter Trško
-- License:      BSD3
--
-- Stability:    experimental
-- Portability:  NoImplicitPrelude
module Data.Word.Length
  where

import Prelude
    ( Integral(quot)
    , Num((+))
    , fromIntegral
    )

import Data.Bool (otherwise)
import Data.Int (Int)
import Data.Ord (Ord((<), (>=)))
import Data.Word (Word, Word16, Word32, Word64, Word8)

import Data.NumberLength.Internal (either32or64)


-- {{{ Decimal ----------------------------------------------------------------

lengthWord8 :: Word8 -> Int
lengthWord8 n   -- Maximum is 255.
  | n < 10     = 1
  | n < 100    = 2
  | otherwise  = 3
{-# INLINE lengthWord8 #-}

lengthWord16 :: Word16 -> Int
lengthWord16 n   -- Maximum is 65535.
  | n < 10     = 1
  | n < 100    = 2
  | n < 1000   = 3
  | n < 10000  = 4
  | otherwise  = 5
{-# INLINE lengthWord16 #-}

lengthWord32 :: Word32 -> Int
lengthWord32 n   -- Maximum is 4294967295.
  | n < 10         = 1
  | n < 100        = 2
  | n < 1000       = 3
  | n < 10000      = 4
  | n >= 100000000 = 8 + lengthWord32 (n `quot` 100000000)
  | otherwise      = 4 + lengthWord32 (n `quot` 10000)
    -- n >= 10000
{-# INLINE lengthWord32 #-}

lengthWord64 :: Word64 -> Int
lengthWord64 n   -- Maximum is 18446744073709551615.
  | n < 10                 = 1
  | n < 100                = 2
  | n < 1000               = 3
  | n < 10000              = 4
  | n >= 10000000000000000 = 16 + lengthWord64 (n `quot` 10000000000000000)
  | n >= 100000000         = 8  + lengthWord64 (n `quot` 100000000)
  | otherwise              = 4  + lengthWord64 (n `quot` 10000)
    -- n >= 10000
{-# INLINE lengthWord64 #-}

lengthWord :: Word -> Int
lengthWord n =
    lengthWord32 (fromIntegral n) `either32or64` lengthWord64 (fromIntegral n)
{-# INLINE lengthWord #-}

-- }}} Decimal ----------------------------------------------------------------

-- {{{ Hexadecimal ------------------------------------------------------------

lengthWord8hex :: Word8 -> Int
lengthWord8hex n    -- Maximum is 255 = 0xff.
  | n < 16    = 1
  | otherwise = 2
{-# INLINE lengthWord8hex #-}

lengthWord16hex :: Word16 -> Int
lengthWord16hex n   -- Maximum is 65535 = 0xffff.
  | n < 0x10   = 1
  | n < 0x100  = 2
  | n < 0x1000 = 3
  | otherwise  = 4
{-# INLINE lengthWord16hex #-}

lengthWord32hex :: Word32 -> Int
lengthWord32hex n   -- Maximum is 4294967295 = 0xffffffff.
  | n < 0x10    = 1
  | n < 0x100   = 2
  | n < 0x1000  = 3
  | n < 0x10000 = 4
  | otherwise   = 4 + lengthWord32hex (n `quot` 0x10000)
    -- n >= 0x10000
{-# INLINE lengthWord32hex #-}

lengthWord64hex :: Word64 -> Int
lengthWord64hex n   -- Maximum is 18446744073709551615 = 0xffffffffffffffff.
  | n <  0x10        = 1
  | n <  0x100       = 2
  | n <  0x1000      = 3
  | n <  0x10000     = 4
  | n >= 0x100000000 = 8 + lengthWord64hex (n `quot` 0x100000000)
  | otherwise        = 4 + lengthWord64hex (n `quot` 0x10000)
    -- n >= 0x10000
{-# INLINE lengthWord64hex #-}

lengthWordHex :: Word -> Int
lengthWordHex n = lengthWord32hex (fromIntegral n)
    `either32or64` lengthWord64hex (fromIntegral n)
{-# INLINE lengthWordHex #-}

-- }}} Hexadecimal ------------------------------------------------------------


src/Data/NumberLength.hs
{-# LANGUAGE CPP #-}
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE NoImplicitPrelude #-}
-- |
-- Module:       $HEADER$
-- Description:  Polymorphic interface for getting number of digits of a number
--               in decimal or hexadecimal representation.
-- Copyright:    (c) 2015-2016, Peter Trško
-- License:      BSD3
--
-- Stability:    experimental
-- Portability:  DefaultSignatures, NoImplicitPrelude
--
-- Polymorphic interface for getting number of digits of a number in decimal or
-- hexadecimal representation.
module Data.NumberLength
    (
      NumberLength(..)
    , SignedNumberLength(..)
    , BoundedNumberLength(..)
    )
  where

import Prelude(Integer, Num((+)), fromIntegral)

import Data.Int (Int, Int16, Int32, Int64, Int8)
import Data.Ord (Ord((<)))
import Data.Word (Word, Word16, Word32, Word64, Word8)
#ifdef HAVE_NATURAL
import Numeric.Natural (Natural)
#endif

import Data.NumberLength.Int
    ( lengthInt
    , lengthInt16
    , lengthInt16hex
    , lengthInt32
    , lengthInt32hex
    , lengthInt64
    , lengthInt64hex
    , lengthInt8
    , lengthInt8hex
    , lengthIntHex
    )
import Data.NumberLength.Integer (lengthInteger, lengthIntegerHex)
import Data.NumberLength.Internal (either32or64)
#ifdef HAVE_NATURAL
import Data.NumberLength.Natural (lengthNatural, lengthNaturalHex)
#endif
import Data.NumberLength.Word
    ( lengthWord
    , lengthWord16
    , lengthWord16hex
    , lengthWord32
    , lengthWord32hex
    , lengthWord64
    , lengthWord64hex
    , lengthWord8
    , lengthWord8hex
    , lengthWordHex
    )


-- | Get number of digits of a number in base 10 and base 16. Note the
-- following:
--
-- * There is no 'Num' constraint, so that type wrappers aren't forced to
--   provide instance for it. This is because there are things represented
--   using numbers, but they aren't numbers, e.g. telephone numbers.
--
-- * This type class doesn't handle signed numbers, in an intuitive way. See
--   also 'SignedNumberLength'.
--
-- * There is a special class for bounded numbers, see 'BoundedNumberLength',
--   that provides similar functionality as 'Prelude.Bounded', but for number
--   of digits in a number.
class NumberLength a where
    -- | Get number of digits in base 10 for specified number. Note that if
    -- number is signed, then this function will return length of its absolute
    -- value.
    --
    -- >>> numberLength (123 :: Int)
    -- 3
    -- >>> numberLength (-123 :: Int)
    -- 3
    --
    -- See also 'signedNumberLength'.
    numberLength :: a -> Int

    -- | Get number of digits in base 16 for specified number. Note that if
    -- number is signed, then this function will return length of its absolute
    -- value.
    --
    -- >>> numberLengthHex (123 :: Int)  -- 123 = 7b in hex
    -- 2
    -- >>> numberLengthHex (-123 :: Int)
    -- 2
    --
    -- See also 'signedNumberLengthHex'.
    numberLengthHex :: a -> Int

-- | Get number of digits of a signed number in base 10 and base 16.
class NumberLength a => SignedNumberLength a where
    {-# MINIMAL signedNumberLengthHex #-}

    -- | Get number of digits in base 10 for specified number.
    --
    -- >>> signedNumberLength (123 :: Int)
    -- 3
    -- >>> signedNumberLength (-123 :: Int)
    -- 4
    --
    -- Default implementation provided if @a@ has also 'Num' and 'Ord'
    -- instances:
    --
    -- @
    -- 'signedNumberLength' n = signLength + 'numberLength' n
    --   where
    --     signLength = if n < 0 then 1 else 0
    -- @
    signedNumberLength :: a -> Int

    default signedNumberLength :: (Num a, Ord a) => a -> Int
    signedNumberLength n = signLength + numberLength n
      where
        signLength = if n < 0 then 1 else 0

    -- | Get number of digits in base 16 for specified number.
    --
    -- >>> signedNumberLengthHex (123 :: Int)
    -- 2
    -- >>> signedNumberLengthHex (-123 :: Int)
    -- 16
    --
    -- Negative number is shown as ones' complement, e.g. @(-123 :: Int) =
    -- ffffffffffffff85@ on 64 bit platform.
    signedNumberLengthHex :: a -> Int

-- | Get maximum number of digits of a number in base 10 and 16. Minimal number
-- of digits is considered to be always 1, and therefore there is no method for
-- it.
class NumberLength a => BoundedNumberLength a where
    -- | Get maximum number of digits of a number in base 10.
    maxNumberLength :: proxy a -> Int

    -- | Get maximum number of digits of a number in base 16.
    maxNumberLengthHex :: proxy a -> Int

-- {{{ Int* -------------------------------------------------------------------

instance NumberLength Int where
    numberLength = lengthInt
    numberLengthHex = lengthIntHex

instance SignedNumberLength Int where
    signedNumberLengthHex n = numberLengthHex (fromIntegral n :: Word)

instance BoundedNumberLength Int where
    maxNumberLength _ = 10 `either32or64` 19
    maxNumberLengthHex _ = 8 `either32or64` 16

instance NumberLength Int64 where
    numberLength = lengthInt64
    numberLengthHex = lengthInt64hex

instance SignedNumberLength Int64 where
    signedNumberLengthHex n = numberLengthHex (fromIntegral n :: Word64)

instance BoundedNumberLength Int64 where
    maxNumberLength _ = 19
    maxNumberLengthHex _ = 16

instance NumberLength Int32 where
    numberLength = lengthInt32
    numberLengthHex = lengthInt32hex

instance SignedNumberLength Int32 where
    signedNumberLengthHex n = numberLengthHex (fromIntegral n :: Word32)

instance BoundedNumberLength Int32 where
    maxNumberLength _ = 10
    maxNumberLengthHex _ = 8

instance NumberLength Int16 where
    numberLength = lengthInt16
    numberLengthHex = lengthInt16hex

instance SignedNumberLength Int16 where
    signedNumberLengthHex n = numberLengthHex (fromIntegral n :: Word16)

instance BoundedNumberLength Int16 where
    maxNumberLength _ = 5
    maxNumberLengthHex _ = 4

instance NumberLength Int8 where
    numberLength = lengthInt8
    numberLengthHex = lengthInt8hex

instance SignedNumberLength Int8 where
    signedNumberLengthHex n = numberLengthHex (fromIntegral n :: Word8)

instance BoundedNumberLength Int8 where
    maxNumberLength _ = 3
    maxNumberLengthHex _ = 2

-- }}} Int* -------------------------------------------------------------------
-- {{{ Word* ------------------------------------------------------------------

instance NumberLength Word where
    numberLength = lengthWord
    numberLengthHex = lengthWordHex

instance BoundedNumberLength Word where
    maxNumberLength _ = 10 `either32or64` 20
    maxNumberLengthHex _ = 8 `either32or64` 16

instance NumberLength Word64 where
    numberLength = lengthWord64
    numberLengthHex = lengthWord64hex

instance BoundedNumberLength Word64 where
    maxNumberLength _ = 20
    maxNumberLengthHex _ = 16

instance NumberLength Word32 where
    numberLength = lengthWord32
    numberLengthHex = lengthWord32hex

instance BoundedNumberLength Word32 where
    maxNumberLength _ = 10
    maxNumberLengthHex _ = 8

instance NumberLength Word16 where
    numberLength = lengthWord16
    numberLengthHex = lengthWord16hex

instance BoundedNumberLength Word16 where
    maxNumberLength _ = 5
    maxNumberLengthHex _ = 4

instance NumberLength Word8 where
    numberLength = lengthWord8
    numberLengthHex = lengthWord8hex

instance BoundedNumberLength Word8 where
    maxNumberLength _ = 3
    maxNumberLengthHex _ = 2

-- }}} Word* ------------------------------------------------------------------

-- | /Since 0.2.0.0/
instance NumberLength Integer where
    numberLength = lengthInteger
    numberLengthHex = lengthIntegerHex

#ifdef HAVE_NATURAL
-- | /Since 0.2.0.0/
instance NumberLength Natural where
    numberLength = lengthNatural
    numberLengthHex = lengthNaturalHex
#endif
